File: scrapers/pharmeasy.py
"""PharmEasy scraper.

PharmEasy's search results page is server-rendered as a Next.js app:
the full product list lives in `props.pageProps.searchResults[]` inside
the `<script id="__NEXT_DATA__">` blob. No browser execution needed.

The endpoint pattern is:
  https://pharmeasy.in/search/all?name=<query>

Item shape (relevant fields from props.pageProps.searchResults[i]):
    name:                              "Dolo 650Mg Strip Of 15 Tablets"
    slug:                              "dolo-650mg-strip-of-15-tablets-44140"
    salePriceDecimal:                  "24.09"   (effective)
    mrpDecimal:                        "32.12"   (MRP)
    productAvailabilityFlags.isAvailable: bool
    moleculeName:                      "PARACETAMOL / ACETAMINOPHEN"

Product URL: https://pharmeasy.in/online-medicine-order/<slug>

ToS posture (see ../LEGAL.md):
  PharmEasy's robots.txt disallows /search/all*. We rate-limit to
  <= 1 req/sec, fetch only public price data, never republish a downloadable
  dataset, and deep-link every result to the source PharmEasy page so they
  receive the user click. This trade-off is documented and a takedown email
  results in immediate removal.
"""
from __future__ import annotations

import json
import re
import time
from typing import Any

from .base import Offer, PharmacyScraper, ScrapeResult, ms_since
from .match_helpers import pick_best

SEARCH_URL = "https://pharmeasy.in/search/all"
NEXT_DATA_RE = re.compile(
    r'<script id="__NEXT_DATA__"[^>]*>(.+?)</script>',
    re.DOTALL,
)
def _extract_next_data(html: str) -> dict | None:
    m = NEXT_DATA_RE.search(html)
    if not m:
        return None
    try:
        return json.loads(m.group(1))
    except json.JSONDecodeError:
        return None


def _to_float(v: Any) -> float | None:
    if v is None:
        return None
    try:
        return float(v)
    except (TypeError, ValueError):
        return None


class PharmEasyScraper(PharmacyScraper):
    domain = "pharmeasy.in"
    platform_id = "PharmEasy"
    default_delivery_days = 3
    default_return_days = 7

    def scrape(self, medicine: dict) -> ScrapeResult:
        start = time.monotonic()
        url = f"{SEARCH_URL}?name={medicine['name'].replace(' ', '+')}"

        try:
            r = self.get(url)
        except PermissionError as e:
            return ScrapeResult(
                medicine_id=medicine["id"],
                status="blocked",
                offer=None,
                error_message=str(e),
                http_status=None,
                duration_ms=ms_since(start),
                via="httpx",
            )
        except Exception as e:
            return ScrapeResult(
                medicine_id=medicine["id"],
                status="error",
                offer=None,
                error_message=f"{type(e).__name__}: {e}",
                http_status=None,
                duration_ms=ms_since(start),
                via="httpx",
            )

        if r.status_code == 429:
            return ScrapeResult(
                medicine_id=medicine["id"],
                status="rate_limited",
                offer=None,
                error_message=None,
                http_status=429,
                duration_ms=ms_since(start),
                via="httpx",
            )
        if r.status_code != 200:
            return ScrapeResult(
                medicine_id=medicine["id"],
                status="error",
                offer=None,
                error_message=f"HTTP {r.status_code}",
                http_status=r.status_code,
                duration_ms=ms_since(start),
                via="httpx",
            )

        data = _extract_next_data(r.text)
        if data is None:
            return ScrapeResult(
                medicine_id=medicine["id"],
                status="error",
                offer=None,
                error_message="__NEXT_DATA__ not found / invalid JSON",
                http_status=r.status_code,
                duration_ms=ms_since(start),
                via="httpx",
            )

        items = (
            data.get("props", {})
            .get("pageProps", {})
            .get("searchResults")
            or []
        )

        best, score = pick_best(
            items,
            target_name=medicine["name"],
            target_pack=medicine.get("pack"),
            name_of=lambda it: it.get("name") or "",
            # PharmEasy's `manufacturerName` + `packShortName` sometimes contain
            # the pack hint when `name` does not.
            pack_text_of=lambda it: " ".join(
                [
                    it.get("packShortName") or "",
                    it.get("manufacturerName") or "",
                ]
            ),
            threshold=0.50,
        )
        if best is None:
            return ScrapeResult(
                medicine_id=medicine["id"],
                status="not_found",
                offer=None,
                error_message=f"no match >=0.50 (top={score:.2f}, items={len(items)})",
                http_status=r.status_code,
                duration_ms=ms_since(start),
                via="httpx",
            )

        sale = _to_float(best.get("salePriceDecimal"))
        mrp = _to_float(best.get("mrpDecimal"))
        if sale is None or mrp is None:
            return ScrapeResult(
                medicine_id=medicine["id"],
                status="error",
                offer=None,
                error_message="missing salePriceDecimal or mrpDecimal",
                http_status=r.status_code,
                duration_ms=ms_since(start),
                via="httpx",
            )

        availability = (best.get("productAvailabilityFlags") or {}).get("isAvailable", False)
        slug = best.get("slug") or ""

        offer = Offer(
            pharmacy=self.platform_id,
            medicine_id=medicine["id"],
            price=sale,
            mrp=mrp,
            delivery_days=self.default_delivery_days,
            in_stock=bool(availability),
            return_days=self.default_return_days,
            url=(
                f"https://{self.domain}/online-medicine-order/{slug}"
                if slug
                else f"https://{self.domain}/"
            ),
        )
        return ScrapeResult(
            medicine_id=medicine["id"],
            status="success",
            offer=offer,
            error_message=None,
            http_status=r.status_code,
            duration_ms=ms_since(start),
            via="httpx",
        )

File: lib/scrapers/pharmeasy.ts
/**
 * PharmEasy TS scraper (mirrors `scrapers/pharmeasy.py`).
 *
 * Fetches the SSR search page and parses `__NEXT_DATA__` for the
 * `props.pageProps.productList` array. (2026-06: PharmEasy renamed
 * `searchResults` -> `productList`, `manufacturerName` -> `manufacturer`,
 * `packShortName` -> `packform`.) See LEGAL.md for the robots.txt
 * posture on this endpoint.
 */

import { pickBest } from "./match";
import type { Offer, ScrapeResult } from "./types";

const SEARCH_URL = "https://pharmeasy.in/search/all";
const NEXT_DATA_RE = /<script id="__NEXT_DATA__"[^>]*>([\s\S]+?)<\/script>/;
const CHROME_UA =
  "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/131.0.0.0 Safari/537.36";

interface PharmEasyItem {
  name?: string;
  slug?: string;
  packform?: string;
  manufacturer?: string;
  moleculeName?: string;
  salePriceDecimal?: string;
  mrpDecimal?: string;
  productAvailabilityFlags?: {
    isAvailable?: boolean;
  };
}

interface PharmEasyData {
  props?: {
    pageProps?: {
      productList?: PharmEasyItem[];
    };
  };
}

function toFloat(v: unknown): number | null {
  if (v === null || v === undefined) return null;
  const n = typeof v === "number" ? v : parseFloat(String(v));
  return Number.isFinite(n) ? n : null;
}

export async function scrapePharmEasy(query: string, pack?: string | null): Promise<ScrapeResult> {
  const start = Date.now();
  const url = `${SEARCH_URL}?name=${query.replace(/ /g, "+")}`;

  let resp: Response;
  try {
    resp = await fetch(url, {
      headers: {
        "User-Agent": CHROME_UA,
        Accept: "text/html,application/xhtml+xml",
        "Accept-Language": "en-IN,en;q=0.9",
        Referer: "https://pharmeasy.in/",
      },
      signal: AbortSignal.timeout(5_500),
    });
  } catch (err) {
    return {
      pharmacy: "PharmEasy",
      status: "error",
      offer: null,
      error_message: err instanceof Error ? err.message : String(err),
      duration_ms: Date.now() - start,
    };
  }

  if (resp.status === 429) {
    return {
      pharmacy: "PharmEasy",
      status: "rate_limited",
      offer: null,
      error_message: null,
      duration_ms: Date.now() - start,
    };
  }
  if (!resp.ok) {
    return {
      pharmacy: "PharmEasy",
      status: "error",
      offer: null,
      error_message: `HTTP ${resp.status}`,
      duration_ms: Date.now() - start,
    };
  }

  const html = await resp.text();
  const m = NEXT_DATA_RE.exec(html);
  if (!m) {
    return {
      pharmacy: "PharmEasy",
      status: "error",
      offer: null,
      error_message: "__NEXT_DATA__ not found",
      duration_ms: Date.now() - start,
    };
  }

  let data: PharmEasyData;
  try {
    data = JSON.parse(m[1]) as PharmEasyData;
  } catch (err) {
    return {
      pharmacy: "PharmEasy",
      status: "error",
      offer: null,
      error_message: `json decode: ${err instanceof Error ? err.message : err}`,
      duration_ms: Date.now() - start,
    };
  }

  const items = data.props?.pageProps?.productList ?? [];
  const { item, score } = pickBest<PharmEasyItem>({
    items,
    targetName: query,
    targetPack: pack,
    nameOf: (it) => it.name ?? "",
    packTextOf: (it) => `${it.packform ?? ""} ${it.manufacturer ?? ""}`,
    threshold: 0.5,
  });
  if (!item) {
    return {
      pharmacy: "PharmEasy",
      status: "not_found",
      offer: null,
      error_message: `no match >=0.50 (top=${score.toFixed(2)}, items=${items.length})`,
      duration_ms: Date.now() - start,
    };
  }

  const sale = toFloat(item.salePriceDecimal);
  const mrp = toFloat(item.mrpDecimal);
  if (sale === null || mrp === null) {
    return {
      pharmacy: "PharmEasy",
      status: "error",
      offer: null,
      error_message: "missing salePriceDecimal / mrpDecimal",
      duration_ms: Date.now() - start,
    };
  }

  const slug = item.slug ?? "";
  const offer: Offer = {
    pharmacy: "PharmEasy",
    medicine_id: "",
    price: sale,
    mrp,
    delivery_days: 3,
    in_stock: Boolean(item.productAvailabilityFlags?.isAvailable),
    return_days: 7,
    url: slug ? `https://pharmeasy.in/online-medicine-order/${slug}` : "https://pharmeasy.in/",
  };

  return {
    pharmacy: "PharmEasy",
    status: "success",
    offer,
    candidate: {
      name: item.name ?? "",
      composition: item.moleculeName ?? null,
      manufacturer: item.manufacturer ?? null,
      pack: item.packform ?? null,
    },
    error_message: null,
    duration_ms: Date.now() - start,
  };
}
